livekit-android-sdk/src/main/java/io/livekit/android/room/track/video/VideoSinkVisibility.kt

```diff
@@ -0,0 +1,136 @@
+package io.livekit.android.room.track.video
+
+import android.graphics.Rect
+import android.os.Handler
+import android.os.Looper
+import android.view.View
+import android.view.ViewTreeObserver
+import androidx.annotation.CallSuper
+import androidx.compose.ui.layout.LayoutCoordinates
+import io.livekit.android.room.track.Track
+import java.util.*
+
+abstract class VideoSinkVisibility : Observable() {
+    abstract fun isVisible(): Boolean
+    abstract fun size(): Track.Dimensions
+
+    /**
+     * This should be called whenever the visibility or size has changed.
+     */
+    fun notifyChanged() {
+        setChanged()
+        notifyObservers()
+    }
+
+    /**
+     * Called when this object is no longer needed and should clean up any unused resources.
+     */
+    @CallSuper
+    open fun close() {
+        deleteObservers()
+    }
+}
+
+class ComposeVisibility : VideoSinkVisibility() {
+    private var lastCoordinates: LayoutCoordinates? = null
+
+    override fun isVisible(): Boolean {
+        return (lastCoordinates?.isAttached == true && lastCoordinates?.size?.width != 0 && lastCoordinates?.size?.height != 0)
+    }
+
+    override fun size(): Track.Dimensions {
+        val width = lastCoordinates?.size?.width ?: 0
+        val height = lastCoordinates?.size?.height ?: 0
+        return Track.Dimensions(width, height)
+    }
+
+    fun onGloballyPositioned(layoutCoordinates: LayoutCoordinates) {
+        val lastVisible = isVisible()
+        val lastSize = size()
+        lastCoordinates = layoutCoordinates
+
+        if (lastVisible != isVisible() || lastSize != size()) {
```

Comment:
Need some Kotlin education :-) or a new pair of glasses :-)

A couple of lines above these variables (`lastVisible` and `lastSize`) are set to return values from the functions checked against here. Under what conditions will those be not equal?

Reply:
the underlying calculations of `isVisible` and `size` depend on the `lastCoordinates`, which gets updated in line 50. They'll be different if the new coordinates are different from the old coordinates.

Reply:
Ah got it. Thank you